lib/adapters/http.js

```diff
@@ -116,16 +116,40 @@ module.exports = function httpAdapter(config) {
     }
 
     if (proxy) {
-      options.hostname = proxy.host;
-      options.host = proxy.host;
-      options.headers.host = parsed.hostname + (parsed.port ? ':' + parsed.port : '');
-      options.port = proxy.port;
-      options.path = protocol + '//' + parsed.hostname + (parsed.port ? ':' + parsed.port : '') + options.path;
-
-      // Basic proxy authorization
-      if (proxy.auth) {
-        var base64 = new Buffer(proxy.auth.username + ':' + proxy.auth.password, 'utf8').toString('base64');
-        options.headers['Proxy-Authorization'] = 'Basic ' + base64;
+      if (!isHttps) {
+        options.hostname = proxy.host;
+        options.host = proxy.host;
+        options.headers.host = parsed.hostname + (parsed.port ? ':' + parsed.port : '');
+        options.port = proxy.port;
+        options.path = protocol + '//' + parsed.hostname + (parsed.port ? ':' + parsed.port : '') + options.path;
+
+        // Basic proxy authorization
+        if (proxy.auth) {
+          var base64 = new Buffer(proxy.auth.username + ':' + proxy.auth.password, 'utf8').toString('base64');
+          options.headers['Proxy-Authorization'] = 'Basic ' + base64;
+        }
+      } else { // use tunnel for https request, referring to https://github.com/request/request
+        var proxyHeaders = Object.assign({}, options.headers);
+        var proxyProtocol = (proxy.protocol === 'https:' ? 'Https' : 'Http');
+        var tunnelFn = require('tunnel-agent')['httpsOver' + proxyProtocol];
+        options.agent = tunnelFn({
+          proxy: {
+            host: proxy.hostname,
+            port: +proxy.port,
+            proxyAuth: proxy.auth,
+            headers: proxyHeaders
+          },
+          headers: options.headers,
+          ca: options.ca,
+          cert: options.cert,
+          key: options.key,
+          passphrase: options.passphrase,
+          pfx: options.pfx,
+          ciphers: options.ciphers,
+          rejectUnauthorized: options.rejectUnauthorized,
+          secureOptions: options.secureOptions,
+          secureProtocol: options.secureProtocol
+        });
```

Comment:
this adapter is pretty big, sort of tricky to follow the logic, would you mind breaking the logic down into methods? my main concern is having deeply nested `if` conditionals, it's sort of hard to ensure our logic is sound with these